lib/heuristics/scheduling_heuristic.rb

```diff
@@ -299,34 +298,34 @@ def update_route(full_route, first_index, first_start = nil)
         previous_end = stop[:end]
       }
 
-      raise OptimizerWrapper::SchedulingHeuristicError, 'Vehicle end violated after updating route' if day_route.size.positive? &&
-                                                                                                       day_route.last[:end] + matrix(full_route, day_route.last[:point_id], full_route[:end_point_id]) > full_route[:tw_end]
+      raise OptimizerWrapper::SchedulingHeuristicError, 'Vehicle end violated after updating route' if route.size.positive? &&
+                                                                                                       route.last[:end] + matrix(route_data, route.last[:point_id], route_data[:end_point_id]) > route_data[:tw_end]
 
-      day_route
+      route
     end
 
-    def collect_unassigned(vrp)
+    def collect_unassigned
       unassigned = []
 
-      @candidate_services_ids.each{ |point|
-        service_in_vrp = vrp.services.find{ |service| service[:id] == point }
+      @candidate_services_ids.each{ |id|
+        service_in_vrp = @services_data[id][:raw]
         (1..service_in_vrp.visits_number).each{ |index|
-          unassigned << get_unassigned_info(vrp, "#{point}_#{index}_#{service_in_vrp.visits_number}", service_in_vrp, 'Heuristic could not affect this service before all vehicles are full')
+          unassigned << get_unassigned_info("#{id}_#{index}_#{service_in_vrp.visits_number}", service_in_vrp, 'Heuristic could not affect this service before all vehicles are full')
         }
       }
 
       @uninserted.each_key{ |service|
-        s = @uninserted[service][:original_service]
-        service_in_vrp = vrp.services.find{ |current_service| current_service[:id] == s }
-        unassigned << get_unassigned_info(vrp, service, service_in_vrp, @uninserted[service][:reason])
+        id = @uninserted[service][:original_service]
+        service_in_vrp = @services_data[id][:raw]
+        unassigned << get_unassigned_info(service, service_in_vrp, @uninserted[service][:reason])
       }
 
       unassigned
     end
 
     def provide_group_tws(services, day)
       services.each{ |service|
-        next if !@services_unlocked_by.collect{ |id, set| [id, set] }.flatten.include?(service.id) || # not in a same_point_day_relation
+        next if !@services_unlocked_by.flat_map{ |id, set| [id, set].flatten }.include?(service.id) || # not in a same_point_day_relation
```

Comment:
Instead of repeatedly creating this `flat_map` for each service can't we create it before the loop and use it inside the loop (even better if this function is called many times during the time where the array `@services_unlocked_by.flat_map{ |id, set| [id, set].flatten }` doesn't change we can create it outside and keep it in the data-structure of the algo)

Reply:
@fonsecadeline  did you forget this one or is there something wrong with what I suggested?